### PR TITLE
Disallow if_succeeds annotation for contract definitions

### PR DESCRIPTION
**Pull request by blitz-1306**
## Changes
- [x] Disallowed any annotation types, except `invariant` type on contracts.
- [x] Disallowed any annotation types, except `if_succeeds` type on functions.
- [x] Overhauled `AnnotationExtractor`:
    - Isolated annotation parsing to `makeAnnotationFromMatch()` funtion for a better readability.
    - Introduced `validateAnnotation()` method to check that created annotation is valid for the target. Moved rules from scribble executable module to this method to concentrate logic in a single place.
    - Tweaked `findAnnotations()` to respect filtering in-place and do not trigger validation on filtered out annotations.
- [x] Introduced `AnnotationError` class. The `SyntaxError` is now based on `AnnotationError`. Also introduced `UnsupportedByTargetError` class for validation errors.
- [x] Minor cleanup in scribble executable module to adapt to a new logic.
- [x] Fixed affected test samples.
- [x] Introduced test cases to check that errors are actually thrown.

Regards.

### CODE REVIEW

**Comment by blitz-1306** [comment]
@cd1m0 So, should we extend the implementation to check invariants over functions? It seems that it is related to https://github.com/ConsenSys/scribble/issues/6#issuecomment-744537926.

**Comment by cd1m0** [comment]
Yes! We should definitely extend the check to make sure that you can't specify an `invariant` annotation on a function

**Review by cd1m0 — APPROVED** [review]
The changes so far look good. Just need to add a similar check that `invariant` doesn't show up on function annotations.

**Comment by blitz-1306** [comment]
@cd1m0 I bit reworked the `AnnotationExtractor` in attemp to concentrate validation rules in one place. Also the changes (in my opinion) will make flow a bit more clear. Here is the simplified flow for annotation extraction:
- Find match in structured documentation.
- Proceed with annotation creation (delimiter check, syntax check and parse predicate expression).
- Check if annotation is passing filter rules. If it not passing, then skip further steps.
- Validate annotation.
- Push annotation to the result array.

We can also consider as an option to filtering prior to annotation parsing to not do additional work for the filtered out annotations. However for now it may be better to do syntax checking at least. So I leave the decision up to you on this.

Regards.

File: src/bin/scribble.ts [hunk 1]
```diff
@@ -267,25 +264,14 @@ function instrumentFiles(
...
-
-                prettyError(
-                    "UnsupportedError",
-                    "Instrumenting free functions is not supported yet",
-                    unit,
-                    annotation.predicateFileLoc(contents),
-                    annotation.original
-                );
-            }
+            getAnnotationsOrDie(fun, ctx.files, filters);
```
Comment (cd1m0) [inline src/bin/scribble.ts:267]: Should add a comment here to explain the logic:

// We call getAnnotationsOrDie to make sure there are no annotations on free functions

File: src/instrumenter/annotations.ts [hunk 2]
```diff
@@ -137,161 +149,210 @@ export class SyntaxError extends Error {
...
+        }
+    }
 
-                    throw new SyntaxError(e.message, original, errRange);
-                }
+    private validateAnnotation(target: AnnotationTarget, annotation: Annotation, source: string) {
+        if (target instanceof ContractDefinition) {
+            if (annotation.type !== AnnotationType.Invariant) {
+                throw new UnsupportedByTargetError(
+                    `Annotation type "${annotation.type}" is not applicable to a target`,
```
Comment (cd1m0) [inline src/instrumenter/annotations.ts:276]: Small rewording:

`"${annotation.type}" annotation is not applicable to contracts`

File: src/instrumenter/annotations.ts [hunk 3]
```diff
@@ -137,161 +149,210 @@ export class SyntaxError extends Error {
...
+                throw new UnsupportedByTargetError(
+                    `Unsupported contract annotations on ${target.kind} ${target.name}`,
+                    annotation.original,
+                    annotation.annotationFileLoc(source)
+                );
+            }
+        } else if (target instanceof FunctionDefinition) {
+            if (annotation.type !== AnnotationType.IfSucceeds) {
+                throw new UnsupportedByTargetError(
+                    `Annotation type "${annotation.type}" is not applicable to a target`,
```
Comment (cd1m0) [inline src/instrumenter/annotations.ts:292]: Small rewording:

"${annotation.type}" annotation is not applicable to functions

**Review by cd1m0 — APPROVED** [review]
Awesome work! 2-3 small re-wordings. Will fix them right now.

File: src/instrumenter/annotations.ts (hunk 3 above)
Comment (blitz-1306) [reply src/instrumenter/annotations.ts]: Addressed in https://github.com/ConsenSys/scribble/pull/5/commits/3b1a65c0b0e464efe1aaada050bfe36bd5592678.

File: src/instrumenter/annotations.ts (hunk 2 above)
Comment (blitz-1306) [reply src/instrumenter/annotations.ts]: Addressed in https://github.com/ConsenSys/scribble/pull/5/commits/3b1a65c0b0e464efe1aaada050bfe36bd5592678.

File: src/bin/scribble.ts (hunk 1 above)
Comment (blitz-1306) [reply src/bin/scribble.ts:270]: Addressed in https://github.com/ConsenSys/scribble/pull/5/commits/3b1a65c0b0e464efe1aaada050bfe36bd5592678.

**Review by cd1m0 — APPROVED** [review]
LGTM

**Pull request merged**